sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala

```diff
@@ -80,4 +80,10 @@ trait Encoder[T] extends Serializable {
    * A ClassTag that can be used to construct an Array to contain a collection of `T`.
    */
   def clsTag: ClassTag[T]
+
+  /**
+   * Create a copied [[Encoder]]. The implementation may just copy internal reusable fields to speed
+   * up the [[Encoder]] creation.
+   */
+  def copyEncoder: Encoder[T]
```

Comment:
How about `makeCopy`? Looks like `TreeNode.makeCopy` tried to avoid `copy`.

Reply:
Seems good `makeCopy`.

Reply:
@HyukjinKwon `makeCopy` sounds better to me as well. Updated.